src/latency.c

```diff
@@ -490,6 +490,147 @@ sds createLatencyReport(void) {
+ * Empty buckets are not printed.
+ * Everything above 1sec is considered +Inf. */
+void fillCommandCDFResp(client *c, const char* histogram_name, struct hdr_histogram* histogram){
+    addReplyArrayLen(c,5);
+    addReplyBulkCString(c,histogram_name);
+    addReplyBulkCString(c,"calls");
+    addReplyLongLong(c,(long long) histogram->total_count);
+    addReplyBulkCString(c,"histogram");
+    void *replylen = addReplyDeferredLen(c);
+    int samples = 0;
+    struct hdr_iter iter;
+    hdr_iter_log_init(&iter, histogram, 1024,2);
+    int64_t previous_count = 0;
+    while (hdr_iter_next(&iter)) {
+        const int64_t micros = iter.highest_equivalent_value / 1000;
+        const int64_t cumulative_count = iter.cumulative_count;
+        if(cumulative_count > previous_count){
+            addReplyLongLong(c,(long long) micros);
+            addReplyLongLong(c,(long long) cumulative_count);
+            samples+=2;
+        }
+        previous_count = cumulative_count;
+    }
+    setDeferredArrayLen(c,replylen,samples);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies. */
+void latencyAllCommandsFillCDF(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    sds info = sdsempty();
+    while((de = dictNext(di)) != NULL) {
+        char *tmpsafe;
+        cmd = (struct redisCommand *) dictGetVal(de);
+        if (!cmd->latency_histogram)
+            continue;
+        info = fillCommandCDF(info, getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe), cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+    }
+    dictReleaseIterator(di);
+    addReplyVerbatim(c,info,sdslen(info),"txt");
+    sdsfree(info);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies. */
+void latencyCommandFillCDF(client *c) {
+    sds info = sdsempty();
+    for (int j = 2; j < c->argc; j++){
+        struct redisCommand *cmd = dictFetchValue(server.commands, c->argv[j]->ptr);
+        if (cmd == NULL) {
+            addReplyErrorFormat(c,
+                "Invalid command name '%s'", (char*) c->argv[j]->ptr);
+            sdsfree(info);
+            return;
+        }
+        char *tmpsafe;
+        if (!cmd->latency_histogram)
+            continue;
+        info = fillCommandCDF(info, getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe), cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+    }
+    addReplyVerbatim(c,info,sdslen(info),"txt");
+    sdsfree(info);
+}
+
+/* latencyCommand() helper to produce a per command category cumulative distribution of latencies in RESP format. */
+void latencyAllCommandsFillCDFResp(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    void *replylen = addReplyDeferredLen(c);
```

Comment:
this is the outer loop, so the damage of deferring reply is low (one syscall per command)